hq_superset/utils.py

```diff
@@ -139,29 +188,109 @@ def _ensure_schema_created(domain):
             if not engine.dialect.has_schema(engine, schema_name):
                 engine.execute(sqlalchemy.schema.CreateSchema(schema_name))
 
-    def re_eval_roles(self, existing_roles, new_domain_role):
-        # Filter out other domain roles
-        new_domain_roles = [
-            r
-            for r in existing_roles
-            if not r.name.startswith(DOMAIN_PREFIX)
-        ] + [new_domain_role]
-        additional_roles = [
-            self.sm.add_role(r)
-            for r in self.sm.appbuilder.app.config['AUTH_USER_ADDITIONAL_ROLES']
-        ]
-        return new_domain_roles + additional_roles
+    def _ensure_schema_perm_created(self, domain):
+        menu_name = self.sm.get_schema_perm(get_hq_database(), get_schema_name_for_domain(domain))
+        permission = self.sm.find_permission_view_menu(SCHEMA_ACCESS_PERMISSION, menu_name)
 
-    def sync_domain_role(self, domain):
-        # This creates DB schema, role and schema permissions for the domain and
-        #   assigns the role to the current_user
-        self._ensure_schema_created(domain)
-        permission = self._ensure_schema_perm_created(domain)
-        role = self._ensure_domain_role_created(domain)
-        self.sm.add_permission_role(role, permission)
-        current_user.roles = self.re_eval_roles(current_user.roles, role)
-        self.sm.get_session.add(current_user)
-        self.sm.get_session.commit()
+        if not permission:
+            permission = self.sm.add_permission_view_menu(SCHEMA_ACCESS_PERMISSION, menu_name)
+        return permission
+
+    def _ensure_domain_role_created(self, domain):
+        # This inbuilt method creates only if the role doesn't exist.
+        return self.sm.add_role(get_role_name_for_domain(domain))
+
+    def _get_additional_user_roles(self, domain):
+        domain_permissions, platform_roles_names = self._get_domain_access(domain)
+        if self._user_has_no_access(domain_permissions):
+            return []
+
+        additional_roles = []
+        if domain_permissions[CAN_WRITE_PERMISSION]:
+            platform_roles_names.append(
+                'gamma',
```

Comment:
should this be the one with "G"? The one on superset is "Gamma" or "gamma"?
Extracted it to a constant would be helpful to avoid confusion

Reply:
> Extracted it to a constant would be helpful to avoid confusion

Good idea